engine/consensus/matching/core.go

```diff
@@ -562,222 +559,244 @@ func (c *Core) CheckSealing() error {
+		// (ii) there must be at least 2 receipts from _different_ ENs
+		//      committing to the result
+		// comment: we evaluate condition (ii) only if (i) is true
+		if happySealable || emergencySealable {
+			// TODO for sealing Phase 3: optimize resultHasMultipleReceipts calls
+			// During Phase 2: for every result, there is only one incorporatedResult, as
+			// incorporatedResult.IncorporatedBlockID is set to the executed block. Without an
+			// execution fork, there is only one incorporatedResult for each block, and the
+			// following code makes the minimum number of resultHasMultipleReceipts calls.
+			hasMultipleReceipts := c.resultHasMultipleReceipts(incorporatedResult)
+			sealingStatus.SetHasMultipleReceipts(hasMultipleReceipts)
+			if hasMultipleReceipts {
+				// add the result to the results that should be sealed
+				results = append(results, incorporatedResult)
 			}
 		}
+	}
 
-		if matched || emergencySealed {
-			// add the result to the results that should be sealed
-			results = append(results, incorporatedResult)
-		}
+	return results, sealingTracker, nil
+}
 
-		if nextUnsealedIsFinalized {
-			if incorporatedResult.Result.BlockID == nextUnsealed {
-				nextUnsealeds = append(nextUnsealeds, &nextUnsealedResult{
-					BlockID:                       incorporatedResult.Result.BlockID,
-					Height:                        block.Height,
-					ResultID:                      incorporatedResult.Result.ID(),
-					IncorporatedResultID:          incorporatedResult.ID(),
-					TotalChunks:                   len(incorporatedResult.Result.Chunks),
-					FirstUnmatchedChunkIndex:      unmatchedIndex,
-					SufficientApprovalsForSealing: matched,
-					QualifiesForEmergencySealing:  emergencySealed,
-				})
-			}
-		}
+// hasEnoughApprovals implements the HAPPY-PATH SEALING-logic. Details:
+// We match ResultApprovals (from the mempool) to the given incorporatedResult
+// and determine whether sufficient number of approvals are known for each chunk.
+// For each of its chunks, the IncorporatedResult tracks internally the added
+// approvals. Here, we go through the ResultApprovals mempool, check whether
+// the approval is from an authorized Verifiers (at the block which incorporates
+// the result). Approvals from all authorized Verifiers are added to
+// IncorporatedResult (which internally de-duplicates Approvals).
+// Returns:
+// * sealingRecord: a record holding information about the incorporatedResult's sealing status
+// * error:
+//   - NoValidChildBlockError: if the block that incorporates `incorporatedResult` does _not_
+//     have a child yet. Then, the chunk assignment cannot be computed.
+//   - All other errors are unexpected and symptoms of internal bugs, uncovered edge cases,
+//     or a corrupted internal node state. These are all fatal failures.
+func (c *Core) hasEnoughApprovals(incorporatedResult *flow.IncorporatedResult, sealingTracker *sealingtracker.SealingTracker) (*sealingtracker.SealingRecord, error) {
```

Comment:
This function is too complex as it takes in a stateful `sealingTracker` type and returns another `sealingRecord` type.

I would suggest not to depend on those types in order to keep it simple.

This caller seems only care about whether there is enough approvals for the given result, and the `FirstUnmatchedChunkIndex` if not enough. In other words, we don't care about the chunk that had enough approvals, right?

So we could let this function return `(FirstUnmatchedChunkIndex int, hasEnoughApprovals bool)`. 

If we had enough approvals, we return `(-1, true)`; if we the 3rd chunk didn't have enough approval, we return `(2, false)`

Reply:
yea. I have thought about this too. Will refactor it 👍

Reply:
👍 I have refactored `hasEnoughApprovals` to: https://github.com/onflow/flow-go/blob/9fd1bbb811c4baf107508f94ea27ef853f34a41c/engine/consensus/matching/core.go#L639

While I still return the `SealingRecord` for the `incorporatedResult`, I have made several modifications which hopefully address the heart of your concern: 
* removed the stateful `sealingTracker` from the arguments
* While I still return a `SealingRecord`, I have applied significant simplifications: 
   * `SealingRecord`  is now directly generated by `hasEnoughApprovals` (i.e. I removed `SealingTracker` as a factory for  `SealingRecord`s)
   * `SealingRecord`  is never nil 
   * Essentially, the `SealingRecord` holds the values `(FirstUnmatchedChunkIndex int, hasEnoughApprovals bool)` plus some other flags for later sealing checks. 
 
Please let me know whether the code is clear enough with these revisions.